nixos/modules/tasks/auto-upgrade.nix

```diff
@@ -78,11 +88,27 @@ let cfg = config.system.autoUpgrade; in
           HOME = "/root";
         } // config.networking.proxy.envVars;
 
-      path = [ pkgs.gnutar pkgs.xz.bin pkgs.gitMinimal config.nix.package.out ];
+      path = [ pkgs.coreutils pkgs.gnutar pkgs.xz.bin pkgs.gitMinimal config.nix.package.out ];
 
       script = ''
-        ${config.system.build.nixos-rebuild}/bin/nixos-rebuild switch ${toString cfg.flags}
-      '';
+        ${lib.optionalString cfg.allowReboot ''
+          set -euo pipefail
+          T=$(mktemp -d)
+          cd "$T"
+          ${config.system.build.nixos-rebuild}/bin/nixos-rebuild build ${toString cfg.flags}
+          booted="$(readlink /run/booted-system/{initrd,kernel,kernel-modules})"
+          built="$(readlink result/{initrd,kernel,kernel-modules})"
+          cd /tmp
```

Comment:
No need to build=>boot or switch. Just do boot=>reboot or switch

https://github.com/NixOS/nixpkgs/blob/6a63021eba70b8c96b65a42b0547c8ca11d200fb/nixos/modules/installer/tools/nixos-rebuild.sh#L354-L356

```
${nixos-rebuild} boot ${toString cfg.flags}
booted="$(readlink /run/booted-system/{initrd,kernel,kernel-modules})"
built="$(readlink /nix/var/nix/profiles/system/{initrd,kernel,kernel-modules})"
if [ "$booted" = "$built" ]; then
  ${nixos-rebuild} switch ${toString cfg.flags}
else
  /run/current-system/sw/bin/shutdown -r +1  
fi
```
Hardcoding `/nix/var/nix/profiles/system` path isn't nice, but 95% of usecases are covered. I think.

Reply:
If someone figures out a use case where it doesn't work, we can change the module, I guess.